File: 3. Longest Substring Without Repeating Characters/go/main3.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {

	n := len(s)
	var ans int
	m := make(map[byte]int)
	for i, j := 0, 0; j < n; j++ {
		if value, ok := m[s[j]]; ok {
			i = max(value, i)
		}
		ans = max(ans, j-i+1)
		m[s[j]] = j + 1
	}

	return ans
}

func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}

func main() {
	var ans int
	ans = lengthOfLongestSubstring("abcdcca")
	fmt.Print(ans)
}

File: 3. Longest Substring Without Repeating Characters/java/Main3.java
import java.util.HashMap;

public class Main3 {
    public static void main(String[] args) {
        
        String s = "abcdcca";
        int ans = 0;
        ans = lengthOfLongestSubstring(s);
        System.out.println(ans);
    }

    public static int lengthOfLongestSubstring(String s) {
        int n = s.length(), ans = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        for(int i = 0, j = 0; j < n; j++){
            if(map.containsKey(s.charAt(j))){
                i = Math.max(map.get(s.charAt(j)), i);
            }
            ans = Math.max(ans, j - i + 1);
            map.put(s.charAt(j), j + 1);
        }

        return ans;
    }
}
